administration: scheduling-and-retries: add retry waiting time section (#977)
fluent/fluent-bit-docs: administration/scheduling-and-retries.md

If the return status was **OK**, it means it was successfully able to process and flush the data, if it returned an **Error** status, means that an unrecoverable error happened and the engine should not try to flush that data again. If a **Retry** was requested, the _Engine_ will ask the _Scheduler_ to retry to flush that data, the Scheduler will decide how many seconds to wait before that happens.

## Configuring Wait Time for Retry

The Scheduler provides two configuration options called **scheduler.cap** and **scheduler.base** which can be set on service section. 

| Key | Description | Default Value | 
| -- | ------------| --------------| 
| scheduler.cap | Set a maximum retry time in second. The property is supported from v1.8.7. | 2000 | 

> | scheduler.base | Set a base of exponential backoff. The property is supported from v1.8.7. | 5 |

Reviewer: These are also mentioned in other places: 
- https://github.com/fluent/fluent-bit-docs/blob/master/administration/configuring-fluent-bit/classic-mode/configuration-file.md
- https://github.com/fluent/fluent-bit-docs/blob/master/administration/configuring-fluent-bit/yaml/configuration-file.md

@DrewZhang13 I think we need to make sure that any place the user finds these settings, they are directed to this description.  So I think we need a link or update there too.

Author: Good point, will add links in those places for reference

> These two configuration determine the waiting time before retry happen. Fluent Bit use exponential-backoff-and-jitter algorithm to determine the waiting time. These configuration means when the Nth retry needed, scheduler will generate a random number X between `(min(base * (Nth power of 2), cap), cap)` as waiting time for this retry.Basically, the **scheduler.base** decide how fast retry could happen, the **scheduler.cap** decide what's the longest waiting time retry could be. 

Reviewer 2: Revise wording?

>These two configurations determine the waiting time before retry attempts. Fluent Bit uses exponential-backoff-and-jitter algorithm to determine the waiting time. These configurations mean that when the Nth retry is needed, scheduler will generate a random number X between `(min(base * (Nth power of 2), cap), cap)` as the waiting time for this retry. Basically, **scheduler.base** decides how fast retries can happen, while **scheduler.cap** decides what's the longest time each retry can be delayed.

Author: Ack, thanks!

> These two configuration determine the waiting time before retry happen. Fluent Bit use exponential-backoff-and-jitter algorithm to determine the waiting time. These configuration means when the Nth retry needed, scheduler will generate a random number X between `(min(base * (Nth power of 2), cap), cap)` as waiting time for this retry.Basically, the **scheduler.base** decide how fast retry could happen, the **scheduler.cap** decide what's the longest waiting time retry could be. 

Reviewer: Can you give an example calculation to help folks understand? Also I think the formula on its own line is best.

Also I actually do not know what this formula means:

> scheduler will generate a random number X between `(min(base * (Nth power of 2), cap), cap)`

What are the upper and lower bounds for the random number generation? Is the `cap` the upper bound? I think we need a more clear way to write this.

Reviewer: The example below does help

Reviewer: Separate the upper and lower bound same as you do in the example, it makes it clearer

Author: Will separate the formula into independent line and add an example following it. 
Basically, lower bound would be the base, and upper bound will keep increasing with retry time goes until reach to the cap.

For detail explanation on exponential-backoff-and-jitter algorithm, please check this [blog](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/).

### Example

> The following example configure the **scheduler.base** as 5 seconds and **scheduler.cap** as 16 seconds. 

Reviewer 2: The following example configure**s**

Add s to configure

```text
[SERVICE]
    Flush            5
    Daemon           off
    Log_Level        debug
    scheduler.base   5
    scheduler.cap    16
```

The waiting time will be:

| Nth retry | waiting time range (seconds) |
| --- | --- | 
| 1 | (5, 16)   |
| 2 | (5*2, 16) |
| 3 | (16, 16)  |

> | 4 | (16, 16)  |

Reviewer 2: Hmm, are you sure about this? From the blog post you submitted it seems like jitter at least in AWS' context means we will always select a random number between 0 and the current exponentually backed off wait time. So it would be 

0-5
0-10
0-16
0-16

Now I'm not sure if that is what the fluent bit code is doing but I believe that is what the amazon article expects it to do. Please let me know if you have a different understanding.

Author: Yeah you are right, it should be the upper bound keep changing until reach to the maximum, the lower bound will be the base. Which the range should be:
5 - 10
5 - 16
5 - 16
5 - 16

Reviewer 2: The logic looks good now.

## Configuring Wait Time for Retry

The Scheduler provides two configuration options called **scheduler.cap** and **scheduler.base** which can be set on service section. 

| Key | Description | Default Value | 
| -- | ------------| --------------| 
| scheduler.cap | Set a maximum retry time in second. The property is supported from v1.8.7. | 2000 | 
| scheduler.base | Set a base of exponential backoff. The property is supported from v1.8.7. | 5 |

These two configuration determine the waiting time before retry happen. Fluent Bit use exponential-backoff-and-jitter algorithm to determine the waiting time. The waiting time is a random number. 

For the Nth retry, the lower bound of the random number will be:

`base`

The upper bound will be:

`min(base * (Nth power of 2), cap)`

Given a example when base is setup to 3, cap is set up to 30. 

> 1st time retry waiting time: The lower bound will be 3, the upper bound will be 3 * 2 = 6. So the waiting time will be a random number between (3, 6).

Reviewer 3: I'd be tempted to remove the extra text in these bits so just "1st retry:..., 2nd retry:..."

2nd time retry waiting time: the lower bound will be 3, the upper bound will be 3 * (2 * 2) = 12. So the waiting time will be a random number between (3, 12).

3rd time retry waiting time: the lower bound will be 3, the upper bound will be 3 * (2 * 2 * 2) = 24. So the waiting time will be a random number between (3, 24).

4th time retry waiting time: the lower bound will be 3, since 3 * (2 * 2 * 2 * 2) = 48 > 30, the upper bound will be 30. So the waiting time will be a random number between (3, 30).

> Basically, the **scheduler.base** decide the lower bound time retry need to wait, the **scheduler.cap** decide the maximum upper bound time retry need to wait. 

Reviewer 3: ```suggestion
Basically, the **scheduler.base** determines the lower bound of time between each retry and the **scheduler.cap** determines the upper bound.
```

> For detail explanation on exponential-backoff-and-jitter algorithm, please check this [blog](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/).

Reviewer 3: ```suggestion
For a detailed explanation of the exponential backoff and jitter algorithm, please check this [blog](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/).
```

Can we rely on this blog always being there? Or should we incorporate it?

### Example

> The following example configure the **scheduler.base** as 3 seconds and **scheduler.cap** as 30 seconds. 

Reviewer 3: ```suggestion
The following example configures the **scheduler.base** as 3 seconds and **scheduler.cap** as 30 seconds.
```
